approaching_emergency_vehicle_plugin/src/approaching_emergency_vehicle_plugin_node.cpp

```diff
@@ -0,0 +1,533 @@
+        double seconds_since_prev_bsm = (current_bsm_timestamp - tracked_erv_.latest_bsm_timestamp).seconds();
+
+        if(seconds_since_prev_bsm < (1.0 / config_.bsm_processing_frequency)){
+          // Do not process ERV's BSM further since not enough time has passed since its previously processed BSM
+          return;
+        }
+      }
+      else{
+        // Do not process BSM further since it is not for the currently tracked ERV
+        return;
+      }
+    }
+
+    // Get ErvInformation object with information from the BSM if it is from an active ERV that is approaching the ego vehicle
+    boost::optional<ErvInformation> erv_information = getErvInformationFromBsm(std::move(msg));
+    if(!erv_information){
+      // BSM is not from an active ERV that is approaching the ego vehicle
+      return;
+    }
+    else{
+      // Update tracked ERV information since this is an active ERV that is approaching the ego vehicle
+      tracked_erv_ = *erv_information;
+      tracked_erv_.latest_update_time = this->get_clock()->now();
+
+      if(!has_tracked_erv_){
+        has_tracked_erv_ = true;
+      }
+    }
+
+    return;
+  }
+
+  double ApproachingEmergencyVehiclePlugin::getSecondsUntilPassing(lanelet::Optional<lanelet::routing::Route>& erv_future_route, const lanelet::BasicPoint2d& erv_position_in_map, 
+                                                                   const double& erv_current_speed, lanelet::ConstLanelet& intersecting_lanelet){
+
+    // Obtain ego vehicle and ERV distances to the end of the intersecting lanelet so neither vehicle will currently be past that point
+    lanelet::ConstLineString2d intersecting_centerline = lanelet::utils::to2D(intersecting_lanelet.centerline());
+    lanelet::BasicPoint2d intersecting_end_point = intersecting_centerline.back();
+
+    // Get ego vehicle's distance to the beginning of the intersecting lanelet
+    // Note: Distance shall be 0.0 if ego vehicle is currently located within the intersecting lanelet
+    double ego_dist_to_lanelet = wm_->routeTrackPos(intersecting_end_point).downtrack - latest_route_state_.down_track;
+
+    // Set erv_world_model_ route to the erv_future_route
+    lanelet::routing::Route route = std::move(*erv_future_route);
+    carma_wm::LaneletRoutePtr erv_future_route_ptr = std::make_shared<lanelet::routing::Route>(std::move(route));
+    erv_world_model_->setRoute(erv_future_route_ptr);
+
+    // Get downtrack of intersecting lanelet on ERV's route
+    double erv_dist_to_lanelet = erv_world_model_->routeTrackPos(intersecting_end_point).downtrack - erv_world_model_->routeTrackPos(erv_position_in_map).downtrack;
+
+    // Calculate seconds_until_passing and protect against division by zero
+    double delta_speed = erv_current_speed - current_speed_;
+    if(delta_speed == 0.0){
+      delta_speed = epsilon_;
+    }
+    double seconds_until_passing = (erv_dist_to_lanelet - ego_dist_to_lanelet) / delta_speed;
```

Comment:
so if erv is slower than ego veh the seconds become negative. that is considered non-aproaching?

Reply:
Yes correct, I have treated it that way in getErvInformationFromBsm(). But to be more explicit, I have added a note on this in the header file for the function.